doc: ipc: Add detailed protocol specification for ICBMsg backend

doki-nordic opened the pull request:
The details of how the ICBMsg communication is done, are hidden inside the source code. This specification adds details about the ICBMsg protocol.

doki-nordic commented:
Direct link to docs: 
https://builds.zephyrproject.io/zephyr/pr/77229/docs/services/ipc/ipc_service/backends/ipc_service_icbmsg.html#detailed-protocol-specification

dchat-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -81,3 +81,299 @@ Samples
+--------------------------
+
+The ICBMsg uses two shared memory regions, ``rx-region`` for message receiving, and ``tx-region`` for message transmission.
+The regions do not have to be next to each other nor placed in any specification order nor have the same size.
```
```suggestion
The regions do not have to be next to each other nor placed in any specific order nor have the same size.
```

pdunaj replied:
I would add a remark that these areas must be aligned to cache boundary and do not overlap.
...
I went further and see there is a remark regarding cache so  maybe this is not needed.

dchat-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -81,3 +81,299 @@ Samples
+  This area is divided into even-sized blocks aligned to cache boundaries.
+
+The location of each area is calculated to fulfill cache boundary requirements and allow optimal region usage.
+It is calculated using the following algorithm.
```
```suggestion
It is calculated using the following algorithm:
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```suggestion
If the follower receives a bounding message before it starts the bounding procedure on that endpoint, it should store the message for later.
```

grochu commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
I am not sure, just checking and don't want to revolutionize the whole page but shouldn't it be _binding_ instead of _bounding_? Technically, the procedure is starting a connection between endpoints on different cores. Do you have an opinion @divipillai?

divipillai replied:
yeah, agree with you Michal. We should be using binding in this scenario, since we are talking about endpoints. Checked for similar example and in NCS, we use the term binding: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/libraries/zigbee/shell.html#zdo_mgmtbind. 

_The correct term should be "binding message" when referring to messages exchanged during the process of establishing a connection between two endpoints in a network. The process itself should be referred to as "endpoint binding." The term "bounding" is not typically used in this context._ - From an another source.

divipillai replied:
Had discussed offline with @doki-nordic , and agreed to change to binding, where ever applicable. Rest docs looks good.

grochu reviewed (commented):
Please look at a few comments from my side.
In general I would like to discuss the verb usage in this documentation file: Can we use the verb _bind_ instead of _bound_? Just suggesting. Any opinions?

doki-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -81,3 +81,299 @@ Samples
 
 * :ref:`ipc_multi_endpoint_sample`
+
+Detailed Protocol Specification
```
I disagree. The "Detailed Protocol Specification" is for users that need to know low level details about the protocol. The configuration is for all user. They do not need to know how the `rx-region` and `tx-region` is constructed. They just need to assign a memory for them using configuration.

doki-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -81,3 +81,299 @@ Samples
+
+The ICBMsg uses two shared memory regions, ``rx-region`` for message receiving, and ``tx-region`` for message transmission.
+The regions do not have to be next to each other nor placed in any specification order nor have the same size.
+Those regions are interchanged on both sides.
```
No, not "interchangeable". I was trying to say that, for example:

If, on APP core, the `rx-region` is at address 1000 and `tx-region` is at address 5000,
then, on RADIO core, the `rx-region` must be at address 5000, `tx-region` must be at address 1000.

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```suggestion
#. The follower is bound to an endpoint, and it can send data messages using this endpoint.
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```suggestion
#. The initiator is bound to an endpoint, and it can send data messages using this endpoint.
```

doki-nordic committed:
doc: ipc: Add detailed protocol specs for ICBMsg backend

The details of how the ICBMsg communication is done, are hidden
inside the source code. This specification adds details about
the ICBMsg protocol.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>

doki-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icbmsg.rst:
```diff
@@ -81,3 +81,299 @@ Samples
+
+The algorithm:
+
+#. Align region boundaries to cache:
```
Keeping `#.` as discussed off-line.
